### PR TITLE
Refactor for v4

### PR DESCRIPTION
**Pull request by thomasrockhu**
It looks like a lot, but it's not. Basically trying to get more logging in and starting to refactor the method

### CODE REVIEW

File: lib/codecov.rb [hunk 1]
```diff
@@ -6,67 +6,135 @@
...
-  VERSION = '0.1.16'
-  def format(result)
-    net_blockers(:off)
+  VERSION = '0.1.17'
+
+  ### CIs
+  RECOGNIZED_CIS = [
+    APPVEYOR = 'Appveyor CI',
+    AZUREPIPELINES = 'Azure Pipelines',
+    BITBUCKET = 'Bitbucket',
```
Comment (hootener) [inline lib/codecov.rb:15]: I think BUDDY (https://buddy.works) should be in here and is missing?

File: lib/codecov.rb [hunk 2]
```diff
@@ -291,10 +288,48 @@ def format(result)
...
+                                'Accept' => 'application/json'
+                              })
+    req.body = json
+
+    https = Net::HTTP.new(uri.host, uri.port)
+    https.use_ssl = !url.match(/^https/).nil?
+
+    # make request
+    puts "#{uri.path}?#{URI.encode_www_form(params_secret_token)}"
+    response = upload_to_codecov(req, https)
```
Comment (hootener) [inline lib/codecov.rb:332]: I assume this is the v4 magic. Any fallback logic to v2 here?

Comment (thomasrockhu) [reply lib/codecov.rb:332]: This currently still just uses the `v1` upload. Worth adding v2 logic as well? As in, is it worth getting a v2 uploader working, and then v4?

File: lib/codecov.rb (hunk 1 above)
Comment (thomasrockhu) [reply lib/codecov.rb:15]: I'm PRETTY sure we don't support `BUDDY` in this uploader

Comment (hootener) [reply lib/codecov.rb:15]: Yeah I just don't think it was ever added, but the codecov bash uploader does support it:

```

elif [ "$BUDDYBUILD_BRANCH" != "" ];
then
  say "$e==>$x buddybuild detected"
  # http://docs.buddybuild.com/v6/docs/custom-prebuild-and-postbuild-steps
  service="buddybuild"
  branch="$BUDDYBUILD_BRANCH"
  build="$BUDDYBUILD_BUILD_NUMBER"
  build_url="https://dashboard.buddybuild.com/public/apps/$BUDDYBUILD_APP_ID/build/$BUDDYBUILD_BUILD_ID"
  # BUDDYBUILD_TRIGGERED_BY
  if [ "$ddp" = "$(echo ~)/Library/Developer/Xcode/DerivedData" ];
  then
    ddp="/private/tmp/sandbox/${BUDDYBUILD_APP_ID}/bbtest"
  fi
```

It's up to you if you want to include at this time.

File: lib/codecov.rb (hunk 2 above)
Comment (hootener) [reply lib/codecov.rb:332]: No, I don't think it matters, per the spec: https://codecovio.atlassian.net/wiki/spaces/ENG/pages/104660999/Upload+Handler+Migration#v1-and-v2

v1 and v2 are functionally identical, so there's no reason to support v2. The reason I ask about a fallback is that the behavior of the bash uploader will attempt the v4 uploader a handful of times, but ultimately fall back to v2 if it cannot upload via v4.

Thanks to changed to our storage infrastructure I'm not sure how necessary this fallback is anymore, though.

**Pull request merged**